### ISSUE TITLE
【BERT】BERT 类模型词表的特殊Token

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#180]
BERT 类模型指的是“Transformer encoder + MLM/分类范式”这一类架构，不限定分词器。

# 常见特殊 Token

常见的有 `[CLS]`, `[SEP]`, `[UNK]`, `[PAD]`, `[MASK]`, `BOS/EOS` 等。

### [CLS], [SEP]

[CLS] 和 [SEP] 是 Transformer 系列模型里常见的特殊标记（special tokens），主要用在 BERT、RoBERTa、DeBERTa 这类模型的输入序列中。tokenizer 把原始文本编码成 subword 序列时，自动在开头和结尾加上的两个特殊 token。具体含义：

1. [CLS]（classification token）
    - 位置：句子最前面；
    - 作用：用于整个序列的聚合表示（全局语义），比如做句子分类、句对分类任务时，经常拿 [CLS] 位置的 hidden state 去接一个分类头；
2. [SEP]（separator token）
    - 位置：句子或句对的分隔位置，一般在单句输入的末尾；
    - 作用：
        - 单句输入：标记句子结束（末尾）；
        - 句对输入：用来分隔两句话（如 [CLS] 文本A [SEP] 文本B [SEP]）；

### [UNK]

`[UNK]`，含义是 `Unknown token`，表示“这段字符/子词不在词表中，无法正常切分/编码”。当 tokenizer 遇到无法用词表覆盖等字符序列时，用 `[UNK]`占位。
- 影响：模型丢失了该片段的具体信息（只能知道“这里有个未知的东西”），过多会影响效果。

### [PAD]

表示 `padding token`：把不同长度的句子补齐到同一长度，便于组成 batch。通常配合 attention_mask 使用，让模型忽略 padding 部分。

### [MASK]

表示 `mask token`：预训练（MLM，Masked Language Modeling）时用，把部分 token 替换成 [MASK] 让模型预测它。微调分类时一般不主动用，但词表里会保留。

### 额外的“保留位/未使用位”

例如 BERT 词表里常见：[unused0] ... [unused99]，作用：给后续扩展词表或特殊标记预留位置（有些项目会拿它们当业务自定义 token 用）。

### BOS/EOS 类（某些模型里）

有些模型不用 [CLS] / [SEP] 命名，而用：
- `<s>`（BOS，begin-of-sequence）
- `</s>`（EOS，end-of-sequence）
例如 RoBERTa、XLM-R 常见。概念上分别对应 “序列开始/结束或分隔”。

### 其他任务相关特殊 token（看模型而定）

- `<unk>`：等价于 [UNK]，只是命名不同
- `<pad>`：等价于 [PAD]
- 还可能有 <mask>、<cls> 等（同理是命名差异）

# 查看模型词表

以 mdeberta-v3-base 为例，mdeberta-v3-base 的 tokenizer_config.json 的内容是

```
{
  "do_lower_case": false,
  "vocab_type": "spm"
}
```

“spm” 说明使用的词表是 SentencePiece，对这种 tokenizer，词表内容不在 tokenizer.json（fast tokenizer 文件）里，而通常在同目录的 SentencePiece 模型文件里，一般叫：
- spm.model / sentencepiece.bpe.model / tokenizer.model / sentencepiece.model 之一，以及（有时还会有）：added_tokens.json、special_tokens_map.json。

**1. 查看有哪些特殊 token**。

```
from transformers import AutoTokenizer

tok = AutoTokenizer.from_pretrained("./models/mdeberta-v3-base", use_fast=False)
print(tok.special_tokens_map)
print(tok.all_special_tokens)
print(tok.all_special_ids)
```

执行输出你会看到如下输出：

```
{'bos_token': '[CLS]', 'eos_token': '[SEP]', 'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}
['[CLS]', '[SEP]', '[UNK]', '[PAD]', '[MASK]']
[1, 2, 3, 0, 250101]
```

**2. 也可以打印各个 token 及对应的 id**：

```
from transformers import AutoTokenizer

tok = AutoTokenizer.from_pretrained("./models/mdeberta-v3-base", use_fast=False)
print("vocab size:", tok.vocab_size)

# 取前 10 个 id->token
for i in range(10):
    print(i, tok.convert_ids_to_tokens(i))
```

输出如下：

```
vocab_size: 250101
0 [PAD] 0.0
1 [CLS] 0.0
2 [SEP] 0.0
3 [UNK] 0.0
4 <0x00> 0.0
5 <0x01> 0.0
6 <0x02> 0.0
7 <0x03> 0.0
8 <0x04> 0.0
9 <0x05> 0.0
```

**并不是所有 BERT “家族” 模型都用 SentencePiece**：
- 原始 BERT 则是用的 WordPiece，例如 bert-base-uncased，词表文件通常是 vocab.txt，
- RoBERTa / GPT-2 使用的 BPE / Byte-level BPE，词表文件通常是 vocab.json + merges.txt（或 tokenizer.json）
- Unigram / Word-level，一些专用模型或老模型会用更简单的词表方式，在主流 transformers 里较少作为新主流。

更多的 tokenizer 对比，可以参见。[^1]

[^1]: https://huggingface.co/docs/transformers/tokenizer_summary
